File: leet/spiral-matrix-ii.ts
/**
 * https://leetcode.com/problems/spiral-matrix-ii/
 */

export function generateMatrix(n: number): number[][] {
  let result: number[][] = Array(n);
  for (let i = 0; i < n; i++) {
    result[i] = Array(n);
  }

  let top = 0;
  let bottom = n - 1;
  let left = 0;
  let right = n - 1;
  let cellCount = n * n;
  let count = 1;

  while (count <= cellCount) {
    // Traverse right
    for (let i = left; i <= right; i++) {
      result[top][i] = count;
      count++;
    }
    if (count > cellCount) return result;
    top++;

    // Traverse down
    for (let i = top; i <= bottom; i++) {
      result[i][right] = count;
      count++;
    }
    if (count > cellCount) return result;
    right--;

    // Traverse left
    for (let i = right; i >= left; i--) {
      result[bottom][i] = count;
      count++;
    }
    if (count > cellCount) return result;
    bottom--;

    // Traverse up
    for (let i = bottom; i >= top; i--) {
      result[i][left] = count;
      count++;
    }
    left++;
  }

  return result;
}

File: returned/python/today4/spiral_matrix_ii.py
# https://leetcode.com/problems/spiral-matrix-ii/description/
from typing import List

class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        size = n*n
        matrix: List[List[int]] = []
        for i in range(n):
            matrix.append([None] * n)
        
        top, bottom = 0, n-1
        left, right = 0, n-1
        count = 1
        
        while count <= size:
            for col in range(left, right+1):
                matrix[top][col] = count
                count += 1
            if count > size:
                return matrix
            top += 1
            
            for row in range(top, bottom+1):
                matrix[row][right] = count
                count += 1
            if count > size:
                return matrix
            right -= 1

            for col in range(right, left-1, -1):
                matrix[bottom][col] = count
                count += 1
            if count > size:
                return matrix
            bottom -= 1

            for row in range(bottom, top-1, -1):
                matrix[row][left] = count
                count += 1
            left += 1
